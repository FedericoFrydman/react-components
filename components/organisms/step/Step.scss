@import '../../assets/styles/brand.scss';
@import '../../assets/styles/colors.scss';
@import '../../assets/styles/fonts.scss';
@import '../../assets/styles/mixins/breakpoints.scss';

.lsux-step-bar {
  .lsux-step-bar--container {
    overflow: hidden;
    position: relative;
  }
  .lsux-step--button {
    align-items: center;
    display: flex;
    padding: 0;
    text-decoration: none;
    &:hover {
      text-decoration: none;
    }
    &:visited {
      color: var(--b200);
    }
  }

  .step-container {
    display: flex;
    justify-content: center;

    &.mobile {
      background-color: $N300;
      height: 4px;

      &.step-circle--current,
      &.step-circle--complete {
        background-color: var(--brand-color-prim200);
      }
    }
  }

  .step {
    align-items: center;
    display: flex;
    flex-direction: column;
    position: relative;
    text-decoration: none;
    z-index: 1;

    .step-circle {
      align-items: center;
      background-color: $N00;
      border: 1px solid $N700;
      border-radius: 50%;
      color: $N800;
      display: flex;
      height: 36px;
      justify-content: center;
      width: 36px;

      img {
        filter: $filter-default;
      }
      &:hover {
        background-color: $N100;
      }
      &.step-circle--current {
        background-color: $N1000;
        color: $N00;
        display: flex;

        img {
          filter: $filter-white;
        }

        + .step-label p {
          color: $N1000;
          font-weight: 600;
        }
        &:hover {
          border: 2px solid $N700;
        }
      }

      &.step-circle--complete {
        border-color: $G200;
        color: $G200;
        display: flex;

        img {
          filter: $filter-green;
        }
      }
    }

    .step-label {
      color: $N800;
      margin-top: 0.5rem;
      text-align: center;
    }
  }

  .step-sections-container {
    background-color: $N200;
    display: flex;
    justify-content: center;

    .step-section-container {
      align-items: center;
      display: flex;
      flex: 1;
      flex-direction: column;
      height: 92px;
      justify-content: center;
      padding-left: 16px;
      padding-right: 16px;
    }

    .step-section {
      background-color: $N400;
      border-radius: 50px;
      height: 6px;
      width: 100%;

      .step-section--bar {
        background-color: var(--brand-color-prim200);
        border-radius: 50px;
        height: 6px;
      }
    }

    .step-section-text {
      margin-bottom: 16px;
    }
  }

  a.step {
    border: none;
    color: $N1000;
    padding: 0;
    text-decoration: none;

    &:focus {
      outline: none;

      .step-label {
        text-decoration: underline;
      }
    }

    &:hover {
      cursor: pointer;
      outline: none;

      .step-circle {
        background-color: $G10;
        border-width: 2px;
        display: flex;
      }

      .step-label p {
        color: $N800;
        font-weight: 600;
        text-decoration: none;
      }
    }
  }

  .auto-divider-right {
    background-color: $N500;
    flex: 1 1 auto;
    height: 1px;
    margin: 18px 0 0 -52px;

    &.small {
      margin: 18px 0 0 -46px;
    }

    &.blank {
      background-color: rgba(0, 0, 0, 0);
    }
  }

  .auto-divider-left {
    background-color: $N500;
    flex: 1 1 auto;
    height: 1px;
    margin: 18px -52px 0 0;

    &.small {
      margin: 18px -46px 0 0;
    }

    &.blank {
      background-color: rgba(0, 0, 0, 0);
    }
  }
}
