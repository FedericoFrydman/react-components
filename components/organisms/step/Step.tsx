import { forwardRef, HTMLProps, Ref, useEffect, useState } from 'react';
import cn from 'classnames';

import { Button } from '../../molecules/button/Button';
import { Container, ContainerBgColors } from '../../atoms/container/Container';
import { Icon } from '../../atoms/icon/Icon';
import { Text } from '../../atoms/text/Text';
import { useViewportSize } from '../../foundation/utilities';
import './Step.scss';

export type StepSize = 'small' | 'medium' | 'auto';

export interface StepSection {
  name: string;
  percentComplete: number;
  isActive: boolean;
}

export interface StepProps extends Omit<HTMLProps<HTMLDivElement>, 'size' | 'onClick' | 'css'> {
  /**
   * Defines step background color
   */
  background?: ContainerBgColors;
  /**
   * Names of the steps in the process
   */
  stepNames?: string[];
  /**
   * Array of the step numbers that have been completed
   */
  completed?: number[];
  /**
   * Current step
   */
  activeIndex: number;
  /**
   * Array of the step numbers that can be clicked upon
   */
  clickable?: number[];
  /**
   * Size of the step control. Small will tighten the spacing and display an ellipse
   * for the non current elements
   */
  size?: StepSize;
  /**
   * Method to call as steps that are clickable are clicked
   */
  onClick: (index: number) => void;
  /**
   * Choose from 2 style variants. Default: 'bubbles'.
   */
  variant?: 'bubbles' | 'slider' | 'section';
  /**
   * The position of the step counter. Default: 'right'.
   */
  counterPosition?: 'right' | 'left';
  /**
   * The step string
   * Use ${0} to interpolate active step text and ${1} for total steps
   */
  stepText?: string;
  /**
   * Adds back button on slider variant
   */
  backSliderClick?: (event: React.MouseEvent) => void;
  /**
   * Sections on step
   */
  sections?: StepSection[];
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLDivElement>;
}

const StepComponent = (
  {
    background = 'white',
    stepNames,
    completed,
    activeIndex,
    size = 'auto',
    classNames = [],
    clickable = [],
    onClick,
    variant = 'bubbles',
    counterPosition = 'right',
    backSliderClick,
    stepText = 'Step ${0} of ${1}',
    sections,
    ...props
  }: StepProps,
  ref: StepProps['ref'],
) => {
  const [stepWidth, setStepWidth] = useState(140);
  const [blockWidth, setBlockWidth] = useState('140px');
  const [labelWidth, setLabelWidth] = useState(120);
  const [ellipse, setEllipse] = useState(false);
  const [isSmall, setSizeSmall] = useState<string>('');

  useEffect(() => {
    const length = sections?.length > 0 ? sections.length : stepNames.length;

    switch (size) {
      case 'small':
        setStepWidth(60);
        setBlockWidth('60px');
        setLabelWidth(40);
        setSizeSmall('small');
        setEllipse(true);
        break;
      case 'auto':
        setStepWidth(140);
        setBlockWidth(`${100 / length}%`);
        setLabelWidth(120);
        setEllipse(false);
        setSizeSmall('');
        break;
      default:
        setStepWidth(140);
        setBlockWidth('140px');
        setLabelWidth(120);
        setEllipse(false);
        setSizeSmall('');
        break;
    }
  }, [size, setStepWidth, stepNames, sections]);

  const { width } = useViewportSize();
  const breakpoint = 1023;

  const activeStep = activeIndex + 1;
  const isDesktop = width > breakpoint;

  return (
    <div {...props} className={cn('lsux-step-bar', classNames)} ref={ref}>
      {variant === 'section' ? (
        <Container background={background} flexbox classNames={['lsux-step-bar--container']}>
          {sections.map((section, index) => (
            <div
              className={'step-sections-container'}
              key={`step-sections-container-${index}`}
              style={{ backgroundColor: `#ededed${section.isActive ? '80' : ''}`, minWidth: blockWidth }}
            >
              <div className="step-section-container">
                <Text classNames={['step-section-text']} bold={section.isActive} text={section.name} />
                <div className="step-section" style={{ backgroundColor: `#c4c4c4${section.isActive ? '' : '80'}` }}>
                  <div className="step-section--bar" style={{ width: `${section.percentComplete}%` }} />
                </div>
              </div>
            </div>
          ))}
        </Container>
      ) : (
        <Container background={background} flexbox classNames={['lsux-step-bar--container']}>
          {stepNames.map((name, index) => {
            let content;
            let complete;
            if (completed.includes(index) && index != activeIndex) {
              content = <Icon name="interface_check" size="medium-small" />;
              complete = 'step-circle--complete';
            } else {
              content = (index + 1).toString();
            }

            const current = cn({ 'step-circle--current': index === activeIndex });
            const stepLabel = ellipse && index != activeIndex ? '...' : name;
            if (index != activeIndex && clickable.includes(index)) {
              const handleClick = (event: React.MouseEvent<HTMLAnchorElement>): void => {
                event.preventDefault();
                onClick(index);
              };

              return isDesktop && variant !== 'slider' ? (
                <div className="step-container" key={index} style={{ minWidth: blockWidth, width: blockWidth }}>
                  <div className={cn(isSmall, 'auto-divider-left', { blank: index === 0 })}></div>
                  <a className={'step'} onClick={handleClick} style={{ minWidth: stepWidth, width: stepWidth }}>
                    <div className={cn('step-circle', current, complete)}>{content}</div>
                    <div className="step-label" style={{ width: labelWidth }}>
                      <Text text={stepLabel} as="p" textSize="small" />
                    </div>
                  </a>
                  <div className={cn(isSmall, 'auto-divider-right', { blank: index === stepNames.length - 1 })}></div>
                </div>
              ) : (
                <div
                  className={cn('step-container', 'mobile', current, complete)}
                  key={index}
                  style={{ minWidth: blockWidth, width: blockWidth }}
                ></div>
              );
            } else {
              return isDesktop && variant !== 'slider' ? (
                <div className="step-container" key={index} style={{ minWidth: blockWidth, width: blockWidth }}>
                  <div className={cn(isSmall, 'auto-divider-left', { blank: index === 0 })}></div>
                  <div className={'step'} style={{ minWidth: stepWidth, width: stepWidth }}>
                    <div className={cn('step-circle', current, complete)}>{content}</div>
                    <div className="step-label" style={{ width: labelWidth }}>
                      <Text text={stepLabel} as="p" textSize="small" />
                    </div>
                  </div>
                  <div className={cn(isSmall, 'auto-divider-right', { blank: index === stepNames.length - 1 })}></div>
                </div>
              ) : (
                <div
                  className={cn('step-container', 'mobile', current, complete)}
                  key={index}
                  style={{ minWidth: blockWidth, width: blockWidth }}
                ></div>
              );
            }
          })}
        </Container>
      )}
      {!isDesktop || variant === 'slider' ? (
        <Container
          flexbox
          justifyContent={backSliderClick ? 'space-between' : counterPosition === 'right' ? 'flex-end' : 'flex-start'}
          flexDirection={backSliderClick && counterPosition === 'left' ? 'row-reverse' : 'row'}
        >
          {backSliderClick ? (
            <Button
              label="Back"
              classNames={['lsux-step--button']}
              variant="tertiary"
              iconLeft="nav_arrow_left"
              iconColor="BRAND200"
              onClick={backSliderClick}
            />
          ) : null}
          <Container flexbox>
            <Text
              classNames={['lsux-step--indicator']}
              text={stepText.replace('${0}', activeStep.toString()).replace('${1}', stepNames?.length?.toString())}
              as="p"
            />
          </Container>
        </Container>
      ) : null}
    </div>
  );
};

/**
 * Step indicator shows a user's current progress in a multi-step task, and steps remaining till completion.
 *
 * Each step is numbered and pairs with a short label to explain the purpose of the step.
 * Previous steps can be clickable to allow users to make edits to the previously entered information.
 */

export const Step = forwardRef<HTMLDivElement, StepProps>(StepComponent);
