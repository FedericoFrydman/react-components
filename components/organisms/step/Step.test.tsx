import { createRef } from 'react';
import { render, screen } from '@testing-library/react';
import { shallow, ShallowWrapper } from 'enzyme';
import cn from 'classnames';

import { Step, StepProps } from './Step';

describe('Step Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    testId = 'test-rbg';
    wrapper = shallow(
      <Step
        activeIndex={1}
        clickable={[0]}
        completed={[0]}
        counterPosition={'right'}
        stepNames={['One', 'Two', 'Three']}
        onClick={null}
      />,
    );
  });

  test('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the RadioGroup the provided classNames', () => {
          const step = renderRadioGroup({ classNames });

          expect(step).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLDivElement>();

    it('provides a ref to the snackbar', () => {
      expect(renderRadioGroup({ ref })).toEqual(ref.current);
    });
  });

  const renderRadioGroup = (props: Omit<StepProps, 'activeIndex' | 'onClick'> = {}): HTMLElement => {
    render(
      <Step
        {...props}
        activeIndex={1}
        clickable={[0]}
        completed={[0]}
        counterPosition={'right'}
        stepNames={['One', 'Two', 'Three']}
        onClick={null}
        data-testid={testId}
      />,
    );

    return screen.getByTestId(testId);
  };
});
