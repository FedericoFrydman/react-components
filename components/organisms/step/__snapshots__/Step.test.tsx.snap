// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step Component should render the component 1`] = `
<div
  className="lsux-step-bar"
>
  <ForwardRef(ContainerComponent)
    background="white"
    classNames={
      Array [
        "lsux-step-bar--container",
      ]
    }
    flexbox={true}
  >
    <div
      className="step-container"
      key="0"
      style={
        Object {
          "minWidth": "140px",
          "width": "140px",
        }
      }
    >
      <div
        className="auto-divider-left blank"
      />
      <a
        className="step"
        onClick={[Function]}
        style={
          Object {
            "minWidth": 140,
            "width": 140,
          }
        }
      >
        <div
          className="step-circle step-circle--complete"
        >
          <ForwardRef(IconComponent)
            name="interface_check"
            size="medium-small"
          />
        </div>
        <div
          className="step-label"
          style={
            Object {
              "width": 120,
            }
          }
        >
          <ForwardRef(TextComponent)
            as="p"
            text="One"
            textSize="small"
          />
        </div>
      </a>
      <div
        className="auto-divider-right"
      />
    </div>
    <div
      className="step-container"
      key="1"
      style={
        Object {
          "minWidth": "140px",
          "width": "140px",
        }
      }
    >
      <div
        className="auto-divider-left"
      />
      <div
        className="step"
        style={
          Object {
            "minWidth": 140,
            "width": 140,
          }
        }
      >
        <div
          className="step-circle step-circle--current"
        >
          2
        </div>
        <div
          className="step-label"
          style={
            Object {
              "width": 120,
            }
          }
        >
          <ForwardRef(TextComponent)
            as="p"
            text="Two"
            textSize="small"
          />
        </div>
      </div>
      <div
        className="auto-divider-right"
      />
    </div>
    <div
      className="step-container"
      key="2"
      style={
        Object {
          "minWidth": "140px",
          "width": "140px",
        }
      }
    >
      <div
        className="auto-divider-left"
      />
      <div
        className="step"
        style={
          Object {
            "minWidth": 140,
            "width": 140,
          }
        }
      >
        <div
          className="step-circle"
        >
          3
        </div>
        <div
          className="step-label"
          style={
            Object {
              "width": 120,
            }
          }
        >
          <ForwardRef(TextComponent)
            as="p"
            text="Three"
            textSize="small"
          />
        </div>
      </div>
      <div
        className="auto-divider-right blank"
      />
    </div>
  </ForwardRef(ContainerComponent)>
</div>
`;
